### PR TITLE
[Geek Camp] Integrate tracing framework into StarRocks

### LINKED ISSUE #4009: Integrate tracing framework into StarRocks 【Moderate】
#  Requirement Description
Integrate tracing framework into StarRocks, currently open-telemetry seems a good fit: https://github.com/open-telemetry/opentelemetry-cpp
1. add opentelementry thirtparty dependency and build successfully
2. add configs about trace colloctor, and common util/manager classes
3. add some trace for demo purpose, e.g. some stream load related RPCs in BE, should be able to see some trace graphs on tracing tool UI, e.g. jaeger

# Mentor
@decster 

# Registration procedure

**Step 1.** Add a comment when you sign up. The comment can be a brief self-introduction and reason for signing up, including what you do, what you're good at, what you're interested in and your technology stack, and how familiar you are with StarRocks. Example:

> I pick! My name is Johnny. I am from StarRocks and I have been working on database kernel development for 1 year. I am interested in OLAP, good at C++ programming, and have been using StarRocks for 6 months. I chose to solve this issue because I am familiar with execution engine.

**Step 2.** On your WeChat, add the StarRocks community assistant "**_StarRocks-1_**" as a WeChat friend and leave your GitHub ID. The assistant will invite you to a WeChat group based on the issue you picked.

**Step 3.** The mentor will communicate with all the participants who picked this issue and assign the task to the best match. 

# Before you start
1. Read the StarRocks [GitHub workflow](https://github.com/StarRocks/community/blob/main/Contributors/guide/workflow.md) carefully, which provides the configurations of BE and FE and describes how to submit a pull request.
2. To learn more about StarRocks, read [StarRocks documentation](https://github.com/StarRocks/docs/blob/main/TOC.md).

# Pull request requirements

1. After you complete code development, upload the code to the StarRocks code repo as a pull request (PR).

  - PR title format: **[SR GeekCamp] + PR title**
  - Labels ：[Geek Camp 2022/feature]
2. After you submit your PR, comment on the status of your PR submission as requested under the Pinned Issue [[StarRocks geek Camp 2022] ](https://github.com/StarRocks/starrocks/issues/4008)page.

  - Comment format: **"I have finished Issue #" + PR link id**

### PR DESCRIPTION
**Pull request by xlwh**
## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] others

## Which issues of this PR fixes ：

Fixes #4009

### CODE REVIEW

File: thirdparty/build-thirdparty.sh [hunk 1]
```diff
@@ -833,6 +833,19 @@ build_vpack() {
...
 
+# opentelemetry
+build_opentelemetry() {
+    check_if_source_exist $OPENTELEMETRY_SOURCE
+
+    cd $TP_SOURCE_DIR/$OPENTELEMETRY_SOURCE
+    mkdir -p $BUILD_DIR
+    cd $BUILD_DIR
+    rm -rf CMakeCache.txt CMakeFiles/
+    $CMAKE_CMD -DCMAKE_INSTALL_PREFIX=${TP_INSTALL_DIR} -DBUILD_TESTING=OFF -DWITH_JAEGER=ON ..
```
Comment (decster) [inline thirdparty/build-thirdparty.sh:844]: add  -DWITH_EXAMPLES=OFF

File: thirdparty/vars.sh [hunk 2]
```diff
@@ -292,7 +292,13 @@ VPACK_NAME="velocypack-XYZ1.0.tar.gz"
 VPACK_SOURCE="velocypack-XYZ1.0"
 VPACK_MD5SUM="161cbf4c347f6daadacfb749c31842f8"
 
+# open-telemetry
+OPENTELEMETRY_DOWNLOAD="https://github.com/open-telemetry/opentelemetry-cpp/archive/refs/tags/v1.2.0.tar.gz"
+OPENTELEMETRY_NAME=v1.2.0.tar.gz
```
Comment (decster) [inline thirdparty/vars.sh:297]: add some name prefix rather than just version, like other libraries

File: be/src/common/tracer.h [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+#include "opentelemetry/trace/provider.h"
+#include "opentelemetry/sdk/trace/simple_processor.h"
+#include "opentelemetry/sdk/trace/tracer_provider.h"
+
+namespace starrocks {
+
+using Span = opentelemetry::nostd::shared_ptr<opentelemetry::trace::Span>;
+using SpanContext = opentelemetry::trace::SpanContext;
+
+class Tracer {
```
Comment (decster) [inline be/src/common/tracer.h:14]: maybe add an example about how to use this class in class comments

**Comment by decster** [comment]
Also, you can experiment with these options:

```
option(WITH_STL "Whether to use Standard Library for C++ latest features" OFF)
option(WITH_GSL
       "Whether to use Guidelines Support Library for C++ latest features" OFF)

option(WITH_ABSEIL "Whether to use Abseil for C++latest features" OFF)

if(NOT DEFINED CMAKE_CXX_STANDARD)
  if(WITH_STL)
    # Require at least C++17. C++20 is needed to avoid gsl::span
    if(CMAKE_VERSION VERSION_GREATER 3.11.999)
      # Ask for 20, may get anything below
      set(CMAKE_CXX_STANDARD 20)
    else()
      # Ask for 17, may get anything below
      set(CMAKE_CXX_STANDARD 17)
    endif()
  else()
    set(CMAKE_CXX_STANDARD 11)
  endif()
endif()
```

We use c++17, so maybe WITH_STL is OK.

File: be/src/common/tracer.h (hunk 3 above)
Comment (xlwh) [reply be/src/common/tracer.h:14]: > maybe add an example about how to use this class in class comments

Thanks a lot for the review, I will add test cases and usage examples later.

**Comment by xlwh** [comment]
Added related dependent libraries and interface wrapper, and a separate PR is integrated it to BE service.  It seems that the new dependencies are not compiled in the CI test, how should I fix it? thanks!  @decster

**Comment by decster** [comment]
> Added related dependent libraries and interface wrapper, and a separate PR is integrated it to BE service. It seems that the new dependencies are not compiled in the CI test, how should I fix it? thanks! @decster

3rd library changes require QE to setup ci environments, will resolve this.
Also since a new library is added, a license file should be added to licenses-binary

**Review by decster — COMMENTED** [review]
like other filenames, LICENSE-opentelemetry-cpp.txt

**Comment by decster** [comment]
> 3rd library changes require QE to setup ci environments, will resolve this.

Talked to QE, the standard process is to separate this PR into 2 PRs, first to add thirdparty only(this will be verified by QE first), second to add other code, could you please separate this PR? Thanks.

**Comment by xlwh** [comment]
> > 3rd library changes require QE to setup ci environments, will resolve this.
> 
> Talked to QE, the standard process is to separate this PR into 2 PRs, first to add thirdparty only(this will be verified by QE first), second to add other code, could you please separate this PR? Thanks.

OK, thanks!

**Comment by wangruin** [comment]
run starrocks_be_unittest

**Comment by decster** [comment]
run starrocks_be_unittest

**Comment by xlwh** [comment]
> run starrocks_be_unittest

It looks like there is a memory issue unrelated to my change, do I need to rebase?

**Comment by xlwh** [comment]
run starrocks_be_unittest

**Comment by decster** [comment]
run starrocks_be_unittest

**Pull request closed without merging**